PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/lib/link/AzureLinkService.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ */
+
+package com.microsoft.azure.toolkit.lib.link;
+
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.project.Project;
+import com.microsoft.azure.toolkit.intellij.link.LinkComposite;
+import com.microsoft.azure.toolkit.intellij.link.ModuleLinkConfig;
+import com.microsoft.azure.toolkit.intellij.link.base.LinkType;
+import com.microsoft.azure.toolkit.intellij.link.base.ServiceType;
+import com.microsoft.azure.toolkit.intellij.link.mysql.MySQLConnectionUtils;
+import com.microsoft.azure.toolkit.intellij.link.mysql.MySQLLinkConfig;
+import com.microsoft.azure.toolkit.intellij.link.mysql.PasswordConfig;
+import com.microsoft.azure.toolkit.intellij.link.mysql.PasswordDialog;
+import com.microsoft.azure.toolkit.intellij.link.po.BaseServicePO;
+import com.microsoft.azure.toolkit.intellij.link.po.LinkPO;
+import com.microsoft.azure.toolkit.intellij.link.po.ModulePO;
+import com.microsoft.azure.toolkit.intellij.link.po.MySQLServicePO;
+import com.microsoft.intellij.AzureLinkStorage;
+import com.microsoft.intellij.AzureMySQLStorage;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+public class AzureLinkService {
+    private static final AzureLinkService instance = new AzureLinkService();
+
+    public static AzureLinkService getInstance() {
+        return AzureLinkService.instance;
+    }
+
+    private AzureLinkService() {
+
+    }
+
+    public void link(Project project, LinkComposite<MySQLLinkConfig, ModuleLinkConfig> linkComposite) {
+        ModulePO modulePO = createModulePO(linkComposite.getModule());
+        MySQLServicePO resource = createResourcePO(linkComposite.getService());
+        // create link
+        LinkPO linkPO = new LinkPO(resource.getId(), modulePO.getId(), LinkType.SERVICE_WITH_MODULE, linkComposite.getEnvPrefix());
+        // storage mysql
+        AzureMySQLStorage.getStorage().addService(resource);
+        // storage password
+        if (ArrayUtils.isNotEmpty(linkComposite.getService().getPasswordConfig().getPassword())) {
+            String inputPassword = String.valueOf(linkComposite.getService().getPasswordConfig().getPassword());
+            if (MySQLConnectionUtils.connect(resource.getUrl(), resource.getUsername(), inputPassword)) {
```

Comment:
we don't save the password when connecting fails?

Reply:
fixed